From 4453db0fbf8e5be721294bbb7a5055012c93d2ad Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sat, 19 Jul 2014 04:30:55 -0400
Subject: [ahci] Fix broken fallback to single MSI mode

Message-id: <20140719043054.19438.62513.stgit@gir.stowe>
Patchwork-id: 86040
O-Subject: [RHEL7.1 PATCH 78/81] ahci: Fix broken fallback to single MSI mode
Bugzilla: 1110895
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110895
Upstream Status: fc40363b2140f5777c88e67814fac9327bf1ee68

commit fc40363b2140f5777c88e67814fac9327bf1ee68
Author: Alexander Gordeev <agordeev@redhat.com>
Date:   Fri Feb 14 14:27:19 2014 -0700

    ahci: Fix broken fallback to single MSI mode

    Commit 7b92b4f61ec4 ("PCI/MSI: Remove pci_enable_msi_block_auto()")
    introduced a regression: if multiple MSI initialization fails, the code
    falls back to INTx rather than to single MSI.

    Fixes: 7b92b4f61ec4 ("PCI/MSI: Remove pci_enable_msi_block_auto()")
    Signed-off-by: Alexander Gordeev <agordeev@redhat.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
    Acked-by: Tejun Heo <tj@kernel.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/ata/ahci.c b/drivers/ata/ahci.c
index 539c62a..7f526d2 100644
--- a/drivers/ata/ahci.c
+++ b/drivers/ata/ahci.c
@@ -1119,8 +1119,10 @@ int ahci_init_interrupts(struct pci_dev *pdev, unsigned int n_ports,
 
  nvec = rc;
  rc = pci_enable_msi_block(pdev, nvec);
- if (rc)
+ if (rc < 0)
   goto intx;
+ else if (rc > 0)
+  goto single_msi;
 
  return nvec;
 
-- 
1.7.1